apps/emqx_gateway_mqttsn/src/emqx_mqttsn_registry.erl

```diff
@@ -17,64 +17,91 @@
 %% @doc The MQTT-SN Topic Registry
 -module(emqx_mqttsn_registry).
 
--behaviour(gen_server).
-
 -include("emqx_mqttsn.hrl").
--include_lib("emqx/include/logger.hrl").
-
--export([start_link/2]).
-
--export([
-    register_topic/3,
-    unregister_topic/2
-]).
 
 -export([
-    lookup_topic/3,
-    lookup_topic_id/3
+    persist_predefined_topics/1,
+    clear_predefined_topics/1
 ]).
 
-%% gen_server callbacks
 -export([
-    init/1,
-    handle_call/3,
-    handle_cast/2,
-    handle_info/2,
-    terminate/2,
-    code_change/3
+    init/0,
+    reg/2,
+    unreg/2,
+    lookup_topic/2,
+    lookup_topic_id/2
 ]).
 
-%% Internal exports (RPC)
--export([
-    do_register/4
-]).
-
--export([lookup_name/1]).
-
--define(SN_SHARD, emqx_mqttsn_shard).
+-define(PKEY(Id), {mqttsn, predef_topics, Id}).
+-define(PKEY_MAX_PREDEF_ID, {mqttsn, max_predef_topic_id}).
 
--record(state, {tabname, max_predef_topic_id = 0}).
+-type registry() :: #{
+    %% The next topic id to be assigned to new registration
+    next_topic_id := pos_integer(),
+    %% The mapping from topic id to topic name
+    id_to_name := map(),
+    %% The mapping from topic name to topic id
+    name_to_id := map()
+}.
 
--record(emqx_mqttsn_registry, {key, value}).
+%%-----------------------------------------------------------------------------
 
--type registry() :: {Tab :: atom(), RegistryPid :: pid()}.
+-spec persist_predefined_topics(list()) -> ok.
+persist_predefined_topics(PredefTopics) when is_list(PredefTopics) ->
+    MaxPredefId = lists:foldl(
+        fun(#{id := TopicId, topic := TopicName0}, AccId) ->
+            TopicName = iolist_to_binary(TopicName0),
+            persistent_term:put(?PKEY(TopicId), TopicName),
+            persistent_term:put(?PKEY(TopicName), TopicId),
+            case TopicId > AccId of
+                true -> TopicId;
+                false -> AccId
+            end
+        end,
+        0,
+        PredefTopics
+    ),
+    persistent_term:put(?PKEY_MAX_PREDEF_ID, MaxPredefId),
+    ok.
 
-%%-----------------------------------------------------------------------------
+-spec clear_predefined_topics(list()) -> ok.
+clear_predefined_topics(PredefTopics) ->
+    lists:foreach(
+        fun(#{id := TopicId, topic := TopicName0}) ->
+            TopicName = iolist_to_binary(TopicName0),
+            persistent_term:erase(?PKEY(TopicId)),
+            persistent_term:erase(?PKEY(TopicName))
+        end,
+        PredefTopics
+    ),
+    persistent_term:erase(?PKEY_MAX_PREDEF_ID),
+    ok.
 
--spec start_link(atom(), list()) ->
-    ignore
-    | {ok, pid()}
-    | {error, Reason :: term()}.
-start_link(InstaId, PredefTopics) ->
-    gen_server:start_link(?MODULE, [InstaId, PredefTopics], []).
+-spec init() -> registry().
+init() ->
+    #{
+        next_topic_id => persistent_term:get(?PKEY_MAX_PREDEF_ID, 0),
```

Comment:
this makes the config change tricky.
e.g. if there are 10 predefined topics when booted, then there are 11-X registered at runtime,
later, if the sysadmin decided to expand the pre-defined topics list, 11 would have already been taken.
 
I would recommend to preserved 1 - N (N could be 100 or 1024 etc) for predefined, and not allow more than that from configuration.
And the reset (N+1 up to 16#FFFF) is for runtime registration.

Reply:
Nice catch, I will fix it asap